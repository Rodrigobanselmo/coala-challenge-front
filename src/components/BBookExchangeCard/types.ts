import { IBook } from "@/core/interfaces/IBook";
import { IBookExchange } from "@/core/interfaces/IBookExchange";

export type BBookCardProps = {
  bookExchange: IBookExchange;
  buttonLabel?: string;
  onClickButton?: (book: IBookExchange) => Promise<void>;
};
