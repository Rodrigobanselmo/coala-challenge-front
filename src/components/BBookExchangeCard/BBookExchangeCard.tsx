import { useAuth } from "../../core/contexts/AuthContext";
import { Button } from "@mui/material";
import { useState } from "react";
import { CardProfile } from "./components/CardProfile";
import {
  StykedCard,
  StyledAuthors,
  StyledBookBox,
  StyledBookInfo,
  StyledImageBox,
  StyledTitle,
} from "./styles";
import { BBookCardProps } from "./types";

export const BBookExchangeCard = ({
  bookExchange,
  buttonLabel,
  onClickButton,
}: BBookCardProps) => {
  const [isLoading, setIsLoading] = useState(false);
  const { user } = useAuth();

  const handleClickButton = async () => {
    if (!onClickButton || isLoading) return;
    setIsLoading(true);
    onClickButton(bookExchange).finally(() => setIsLoading(false));
  };

  const interestBook = bookExchange.interestBook;
  const askingBook = bookExchange.askingBook;

  return (
    <StykedCard>
      <StyledBookBox>
        <StyledImageBox
          src={interestBook?.thumbnail || "/images/book-placeholder.png"}
          alt={interestBook?.title || "livro-placeholder"}
          height={100}
          width={66}
        />
        <StyledBookInfo>
          <StyledTitle>{interestBook?.title}</StyledTitle>
          <StyledAuthors>{interestBook?.authors?.join(", ")}</StyledAuthors>
        </StyledBookInfo>
      </StyledBookBox>
      <Button sx={{ height: 50, m: "auto" }}>tocar</Button>
      <StyledBookBox>
        {askingBook?.thumbnail && (
          <StyledImageBox
            src={askingBook?.thumbnail || "/images/book-placeholder.png"}
            alt={askingBook?.title || "livro-placeholder"}
            height={100}
            width={66}
          />
        )}
        <CardProfile user={user} />
        <StyledBookInfo>
          <StyledTitle>{askingBook?.title}</StyledTitle>
          <StyledAuthors>{askingBook?.authors?.join(", ")}</StyledAuthors>
        </StyledBookInfo>
      </StyledBookBox>
    </StykedCard>
  );
};
