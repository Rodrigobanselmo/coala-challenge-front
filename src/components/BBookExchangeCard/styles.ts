import { Box, Button, IconButton, Typography, styled } from "@mui/material";
import Image from "next/image";

export const StykedCard = styled(Box)`
  display: flex;
  align-items: stretch;
  padding: 20px;
  overflow-x: hidden;
`;

export const StyledBookBox = styled(Box)`
  display: flex;
  flex: 1;
  padding: 20px;
  border: 1px solid #ccc;
  margin: 10px;
  align-items: center;
  overflow: hidden;
  background-color: ${(props) => props.theme.palette.background.paper};
  border-radius: 8px;
`;

export const StyledBookInfo = styled(Box)`
  flex: 1;
  padding: 0 10px;
  white-space: pre-wrap;
`;

export const StyledImageBox = styled(Image)`
  border-radius: 8px;
  border: 2px solid;
  border-color: ${(props) => props.theme.palette.divider};
  max-width: 100%;

  @media (max-width: 600px) {
    width: 100%;
  }
`;

export const StyledTitle = styled("h2")`
  font-size: 1rem;
`;

export const StyledAuthors = styled(Typography)``;

export const StyledExchangeButton = styled(Button)<{ is_visible?: boolean }>`
  border: none;
  border-radius: 4px;
  padding: 8px 16px;
  cursor: pointer;
  transition: background-color 0.3s, opacity 0.3s;
  margin-top: 16px;
  opacity: ${(props) => (props.is_visible ? "1" : "0")};
  position: absolute;
  bottom: 16px;
  left: 50%;
  width: 80%;
  transform: translateX(-50%);
  font-size: 0.8rem;
`;

export const StyledButton = styled(Button)``;
