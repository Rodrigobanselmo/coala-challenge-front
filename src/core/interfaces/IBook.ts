export type IBook = {
  id: number;
  title: string;
  description: string;
  authors?: string[];
  categories?: string[];
  pageCount?: number;
  googleId?: string;
  publishedAt?: Date;
  language?: string;
  smallThumbnail?: string;
  thumbnail?: string;
  updatedAt: Date;
  createdAt: Date;
  deletedAt?: Date;
};
