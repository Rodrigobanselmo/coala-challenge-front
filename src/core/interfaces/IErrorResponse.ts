import { IconProps } from "@mui/material";

export interface IErrorResponse {
  response: {
    status: number;
    data: {
      message: string;
    };
  };
}
