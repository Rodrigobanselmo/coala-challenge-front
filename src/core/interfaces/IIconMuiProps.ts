import { IconProps } from "@mui/material";

export interface IIconMuiProps {
  muiName?: string;
  fontSize?: IconProps["fontSize"];
  color?: IconProps["color"];
}
