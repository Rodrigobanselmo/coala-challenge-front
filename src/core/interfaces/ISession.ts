import { IUser } from "./IUser";

export type ISession = {
  token: string;
  user: IUser;
};
