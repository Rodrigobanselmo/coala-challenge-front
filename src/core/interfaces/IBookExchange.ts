import { StatusEnum } from "../enums/status.enum";
import { IBook } from "./IBook";
import { IUser } from "./IUser";

export type IBookExchange = {
  id: number;
  interestBookId: number;
  targetUserId: string;
  askingBookId: number;
  interetUserId: string;
  updatedAt: Date;
  createdAt: Date;
  deletedAt: Date | null;
  interestBook?: IBook;
  targetUser?: IUser;
  askingBook?: IBook;
  interetUser?: IUser;
  status: StatusEnum;
};
