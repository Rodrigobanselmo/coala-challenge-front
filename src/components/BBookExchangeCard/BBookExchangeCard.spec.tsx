import React from "react";
import { render } from "@testing-library/react";
import { BBookExchangeCard } from "./BBookExchangeCard";
import { IBookExchange } from "@/core/interfaces/IBookExchange";
import { StatusEnum } from "@/core/enums/status.enum";

const mockBookExchange: IBookExchange = {
  status: StatusEnum.PENDING,
  deletedAt: null,
  id: 1,
  askingBookId: 1,
  interetUserId: "1",
  interestBookId: 1,
  targetUserId: "1",
  createdAt: new Date(),
  updatedAt: new Date(),
  interestBook: {
    title: "Interest Book",
    id: 1,
    createdAt: new Date(),
    updatedAt: new Date(),
    authors: ["Author 1", "Author 2"],
    thumbnail: "/interest-book.png",
  },
  askingBook: {
    id: 1,
    createdAt: new Date(),
    updatedAt: new Date(),
    title: "Asking Book",
    authors: ["Author A", "Author B"],
    thumbnail: "/asking-book.png",
  },
};

jest.mock("../../core/contexts/AuthContext", () => {
  return {
    useAuth: () => ({
      user: {
        id: "1",
        email: "qwe@qwe.com",
      },
    }),
  };
});

describe("BBookExchangeCard", () => {
  it("matches the snapshot", () => {
    const { container } = render(
      <BBookExchangeCard
        bookExchange={mockBookExchange}
        buttonLabel="Test Button"
        onClickButton={async () => {}}
      />
    );

    // Use Jest's snapshot feature to capture the rendered output
    expect(container).toMatchSnapshot();
  });

  it("matches the snapshot with missing askingBook", () => {
    const { container } = render(
      <BBookExchangeCard
        bookExchange={{ ...mockBookExchange, askingBook: undefined }}
        buttonLabel="Test Button"
        onClickButton={async () => {}}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("matches the snapshot with missing interestBook", () => {
    const { container } = render(
      <BBookExchangeCard
        bookExchange={{ ...mockBookExchange, interestBook: undefined }}
        buttonLabel="Test Button"
        onClickButton={async () => {}}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
