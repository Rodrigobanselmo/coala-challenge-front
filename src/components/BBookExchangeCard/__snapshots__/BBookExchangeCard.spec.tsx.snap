// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BBookExchangeCard matches the snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-5w22mz"
  >
    <div
      class="MuiBox-root css-tl6mcv"
    >
      <img
        alt="Interest Book"
        class="css-qgb03h"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=%2Finterest-book.png&w=256&q=75"
        srcset="/_next/image?url=%2Finterest-book.png&w=96&q=75 1x, /_next/image?url=%2Finterest-book.png&w=256&q=75 2x"
        style="color: transparent;"
        width="66"
      />
      <div
        class="MuiBox-root css-j1dvnu"
      >
        <h2
          class="css-11aljab"
        >
          Interest Book
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 css-16gwivt-MuiTypography-root"
        >
          Author 1, Author 2
        </p>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-i90wq5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      tocar
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiBox-root css-tl6mcv"
    >
      <img
        alt="Asking Book"
        class="css-qgb03h"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=%2Fasking-book.png&w=256&q=75"
        srcset="/_next/image?url=%2Fasking-book.png&w=96&q=75 1x, /_next/image?url=%2Fasking-book.png&w=256&q=75 2x"
        style="color: transparent;"
        width="66"
      />
      <div
        class="MuiBox-root css-0"
        style="align-items: center; display: flex;"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-8v8enj-MuiAvatar-root"
          data-testid="BAvatar"
        >
          <img
            alt=""
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="/icons/profile.svg"
          />
        </div>
      </div>
      <div
        class="MuiBox-root css-j1dvnu"
      >
        <h2
          class="css-11aljab"
        >
          Asking Book
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 css-16gwivt-MuiTypography-root"
        >
          Author A, Author B
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`BBookExchangeCard matches the snapshot with missing askingBook 1`] = `
<div>
  <div
    class="MuiBox-root css-5w22mz"
  >
    <div
      class="MuiBox-root css-tl6mcv"
    >
      <img
        alt="Interest Book"
        class="css-qgb03h"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=%2Finterest-book.png&w=256&q=75"
        srcset="/_next/image?url=%2Finterest-book.png&w=96&q=75 1x, /_next/image?url=%2Finterest-book.png&w=256&q=75 2x"
        style="color: transparent;"
        width="66"
      />
      <div
        class="MuiBox-root css-j1dvnu"
      >
        <h2
          class="css-11aljab"
        >
          Interest Book
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 css-16gwivt-MuiTypography-root"
        >
          Author 1, Author 2
        </p>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-i90wq5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      tocar
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiBox-root css-tl6mcv"
    >
      <div
        class="MuiBox-root css-0"
        style="align-items: center; display: flex;"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-8v8enj-MuiAvatar-root"
          data-testid="BAvatar"
        >
          <img
            alt=""
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="/icons/profile.svg"
          />
        </div>
      </div>
      <div
        class="MuiBox-root css-j1dvnu"
      >
        <h2
          class="css-11aljab"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-16gwivt-MuiTypography-root"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`BBookExchangeCard matches the snapshot with missing interestBook 1`] = `
<div>
  <div
    class="MuiBox-root css-5w22mz"
  >
    <div
      class="MuiBox-root css-tl6mcv"
    >
      <img
        alt="livro-placeholder"
        class="css-qgb03h"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=%2Fimages%2Fbook-placeholder.png&w=256&q=75"
        srcset="/_next/image?url=%2Fimages%2Fbook-placeholder.png&w=96&q=75 1x, /_next/image?url=%2Fimages%2Fbook-placeholder.png&w=256&q=75 2x"
        style="color: transparent;"
        width="66"
      />
      <div
        class="MuiBox-root css-j1dvnu"
      >
        <h2
          class="css-11aljab"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-16gwivt-MuiTypography-root"
        />
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-i90wq5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      tocar
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiBox-root css-tl6mcv"
    >
      <img
        alt="Asking Book"
        class="css-qgb03h"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=%2Fasking-book.png&w=256&q=75"
        srcset="/_next/image?url=%2Fasking-book.png&w=96&q=75 1x, /_next/image?url=%2Fasking-book.png&w=256&q=75 2x"
        style="color: transparent;"
        width="66"
      />
      <div
        class="MuiBox-root css-0"
        style="align-items: center; display: flex;"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-8v8enj-MuiAvatar-root"
          data-testid="BAvatar"
        >
          <img
            alt=""
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="/icons/profile.svg"
          />
        </div>
      </div>
      <div
        class="MuiBox-root css-j1dvnu"
      >
        <h2
          class="css-11aljab"
        >
          Asking Book
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 css-16gwivt-MuiTypography-root"
        >
          Author A, Author B
        </p>
      </div>
    </div>
  </div>
</div>
`;
